img2dataset/async_downloader.py

```diff
@@ -0,0 +1,339 @@
+            return (False, row)
+
+    def download_shard(
+        self,
+        row,
+    ):
+        """Function to start an image downloading in one process"""
+
+        shard_id, shard_file = row
+        start_time = time.time()
+
+        fs, shard_path = fsspec.core.url_to_fs(shard_file)
+        with fs.open(shard_path, "rb") as f:
+            df = pa.ipc.open_file(f).read_all()
+        schema = df.schema
+        schema = (
+            schema.append(pa.field("key", pa.string()))
+            .append(pa.field("status", pa.string()))
+            .append(pa.field("error_message", pa.string()))
+            .append(pa.field("width", pa.int32()))
+            .append(pa.field("height", pa.int32()))
+            .append(pa.field("original_width", pa.int32()))
+            .append(pa.field("original_height", pa.int32()))
+        )
+        if self.extract_exif:
+            schema = schema.append(pa.field("exif", pa.string()))
+
+        if self.compute_hash is not None:
+            schema = schema.append(pa.field(self.compute_hash, pa.string()))
+
+        pydict = df.select(self.column_list).to_pydict()
+        shard_to_dl = list(enumerate(zip(*(pydict[col] for col in self.column_list))))
+        del pydict
+        del df
+
+        status_dict = CappedCounter()
+
+        count = len(shard_to_dl)
+        successes = 0
+        failed_to_download = 0
+        failed_to_resize = 0
+        url_indice = self.column_list.index("url")
+        caption_indice = self.column_list.index("caption") if "caption" in self.column_list else None
+        bbox_indice = self.column_list.index(self.blurring_bbox_col) if self.blurring_bbox_col is not None else None
+        key_url_list = [(key, x[url_indice]) for key, x in shard_to_dl]
+
+        # give schema to writer
+        sample_writer = self.sample_writer_class(
+            shard_id,
+            self.output_folder,
+            self.save_caption,
+            self.oom_shard_count,
+            schema,
+            self.encode_format,
+        )
+        oom_sample_per_shard = math.ceil(math.log10(self.number_sample_per_shard))
+
+        async def download_task(data_queue, download_semaphore):
+            timeout = ClientTimeout(total=self.timeout)
+            async with ClientSession(timeout=timeout, headers=make_headers(self.user_agent_token)) as session:
+                all_task = [
+                    download_image_with_retry(
+                        x,
+                        semaphore=download_semaphore,
+                        session=session,
+                        queue=data_queue,
+                        retries=self.retries,
+                        user_agent_token=self.user_agent_token,
+                        disallowed_header_directives=self.disallowed_header_directives,
+                    )
+                    for x in key_url_list
+                ]
+                await asyncio.gather(*all_task)
```

Comment:
I think error handling it not right, I'm getting only "Image decoding error" as error in the table https://wandb.ai/rom1504/img2dataset/runs/uuo8cqpt?workspace=user-rom1504

which seems wrong

Reply:
I expect to see some download error (for example 404)